docs/zed/design.md

```diff
@@ -0,0 +1,483 @@
+partial results instead of deriving the answers from scratch.
+
+When and how such partial aggregations are performed is simply a matter of
+writing Zed queries that take the raw data and produce the derived analytics
+while conforming to a naming model that allows the Zed lake to recognize
+the relationship between the raw data and the derived data.
+
+> TBD: Work out these details which are reminiscent of the analytics cache
+> developed in our earlier prototype.
+
+## Keyless Data
+
+This is TBD.  Data without a key should be accepted some way or another.
+One approach is to simply assign the "zero-value" as the pool key; another
+is to use a configured default value.  This would make key-based
+retention policies more complicated.
+
+Another approach would be to create a sub-pool on demand when the first
+keyless data is encountered, e.g., `pool-name.$nokey` where the pool key
+is configured to be "this".  This way, an app or user could query this pool
+by this name to scan keyless data.
+
+## Relational Model
+
+Since a Zed lake can provide strong consistency, workflows that manipulate
+data in a lake can utilize a model where updates are made to the data
+in place.  Such updates involve creating new commits from the old data
+where the new data is a modified form of the old data.  This provides
+emulation of row-based updates and deletes.
+
+If the pool-key is chosen to be "this" for such a use case, then unique
+rows can be maintained by trivially detected duplicates (because any
+duplicate row will be adjacent when sorted by "this") so that duplicates are
+trivially detected.
+
+Efficient upserts can be accomplished because each data object is sorted by the
+pool key.  Thus, an upsert can be sorted then merge-joined with each
+overlapping object.  Where data objects produce changes and additions, they can
+be forwarded to a streaming add operator and the list of modified objects
+accumulated.  At the end of the operation, then new commit(s) along with
+the to-be-deleted objects can be added to the journal in a single atomic
+operation.  A write conflict occurs if there are any other deletes added to
+the list of to-be-deleted objects.  When this occurs, the transaction can
+simply be restarted.  To avoid inefficiency of many restarts, an upsert can
+be partitioned into smaller objects if the use case allows for it.
+
+> TBD: Finish working through this use case, its requirements, and the
+> mechanisms needed to implement it.  Write conflicts will need to be
+> managed at a layer above the journal or the journal extended with the
+> needed functionality.
+
+
+#### 1.5.3 Type Rule
+
+A type rule indicates that all values of any field of a specified type
+be indexed where the type signature uses Zed type syntax.
+For example,
+```
+zed index create IndexGroupEx type ip
+```
+creates a rule that indexes all IP addresses appearing in fields of type `ip`
+in the index group `IndexGroupEx`.
+
+#### 1.5.4 Aggregation Rule
```

Comment:
Maybe a TBD for adding Field rules here as well?

Reply:
I didn't realize this file was split apart to streamline the README, feel free to disregard. We can revisit.